contracts/test/ERC20Permit.t.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.7;
+
+import { DSTest } from "../../modules/ds-test/src/test.sol";
+
+import { ERC20PermitUser } from "./accounts/ERC20User.sol";
+
+import { MockERC20Permit }  from "./mocks/MockERC20.sol";
+
+import { Hevm }          from "./utils/Hevm.sol";
+import { InvariantTest } from "./utils/InvariantTest.sol";
+
+import { ERC20Test, MockERC20 } from "./ERC20.t.sol";
+
+contract ERC20BaseTest is ERC20Test {
+
+    function setUp() override public {
+        token = MockERC20(address(new MockERC20Permit("Token", "TKN", 18)));
+    }
+
+}
+
+contract ERC20PermitTest is DSTest {
+
+    Hevm hevm;
+
+    MockERC20Permit token;
+    ERC20PermitUser       usr;
+
+    uint256 skOwner   = 1;
+    uint256 skSpender = 2;
+    uint256 nonce     = 0;
+    uint256 deadline  = 5000000000; // Timestamp far in the future
+
+    address owner;
+    address spender;
+
+    uint256 constant WAD = 10 ** 18;
+
+    function setUp() external {
+        hevm = Hevm(address(bytes20(uint160(uint256(keccak256("hevm cheat code"))))));
+
+        owner   = hevm.addr(skOwner);
+        spender = hevm.addr(skSpender);
+
+        hevm.warp(deadline - 52 weeks);
+        token = new MockERC20Permit("Maple Token", "MPL", 18);
+        usr   = new ERC20PermitUser();
+    }
+
+    function test_typehash() external {
+        assertEq(token.PERMIT_TYPEHASH(), keccak256("Permit(address owner,address spender,uint256 amount,uint256 nonce,uint256 deadline)"));
+    }
+
+    function test_domainSeparator() external {
+        assertEq(token.DOMAIN_SEPARATOR(), 0x06c0ee43424d25534e5af6b6af862333b542f6583ff9948b8299442926099eec);
+    }
+
+    function test_permit() external {
+        uint256 amount = 10 * WAD;
+        assertEq(token.nonces(owner),             0);
+        assertEq(token.allowance(owner, spender), 0);
+
+        (uint8 v, bytes32 r, bytes32 s) = getValidPermitSignature(amount, owner, skOwner, deadline);
+        assertTrue(usr.try_erc20_permit(address(token), owner, spender, amount, deadline, v, r, s));
+
+        assertEq(token.allowance(owner, spender), amount);
+        assertEq(token.nonces(owner),             1);
+    }
+
+    function test_permitZeroAddress() external {
+        uint256 amount = 10 * WAD;
+        (uint8 v, bytes32 r, bytes32 s) = getValidPermitSignature(amount, owner, skOwner, deadline);
+        assertTrue(!usr.try_erc20_permit(address(token), address(0), spender, amount, deadline, v, r, s));
```

Comment:
I think we should update this whole repo to forge so that we can use vm.expectRevert here and would just use deep fuzzing instead of `prove_`